[BugFix] Fix the bug of empty table after recover table (#13797)

srlch opened the pull request:
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
Fixes #13797
Problem:
This Problem was introduced by commit #12868. This commit use Collections.unmodifiableSet instead of HashSet to reduce the memory usage of FE. But the problem here is that Collections.unmodifiableSet just returns the view of nameToTable.keySet(). Before using the TableNames in recycle database, nameToTable had been dropped, and also the view of it would become empty, which caused this problem.

Solution:
We copy the nameToTable.keySet() only when dropping the database.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
<!--Checkmate-->
- [x] I have checked the version labels which the pr will be auto backported to target branch

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 9 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|6|50.00%|[482, 483, 2815]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
run starrocks_admit_test

srlch commented:
run #39455

srlch commented:
run 39455

srlch commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 9 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|6|50.00%|[482, 483, 2815]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -423,8 +423,8 @@ public void dropDb(String dbName, boolean isForceDrop) throws DdlException, Meta
-                Set<String> tableNames = db.getTableNamesWithLock();
-                runnableList = unprotectDropDb(db, isForceDrop, false);
+                Set<String> tableNames = new HashSet<String>();
+                runnableList = unprotectDropDb(db, isForceDrop, false, tableNames);
```
You can copy the name to a new set here, instead of add a param in unprotectDropDb.

gengjun-git replied:
Besides, better to change the getTableNamesWithLock to a name that reminds the caller that this is just a view of the tablet map.

srlch replied:
ok,i have fixed it.

srlch committed:
[BugFix] bugfix_get_empty_tableNames_in_dropDb (#13797)

Problem:
This Problem was introduced by commit bdb4da1. This commit use Collections.unmodifiableSet instead of new HashSet to reduce the memory usage of fe. But the problem here is that Collections.unmodifiableSet is just return the view of nameToTable.keySet(). Before use the TableNames in recycleDatabase, nameToTable had been dropped and also the view of it would become empty, which cause this problem.

Solution:
We copy the nameToTable.keySet() only when drop the database.

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 4 (25.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|2|00.00%|[260, 716]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|2|50.00%|[482]|

wanpengfei-git commented:
run starrocks_admit_test

wanpengfei-git commented:
@Mergifyio backport branch-2.5
